client/src/components/practice/long/Typing/index.tsx

```diff
@@ -137,120 +161,57 @@ export default function PracticeLongTyping({ content }: PracticeLongTypingProps)
+      const currComponents = disassemble(value[value.length - 1]);
+      typingInfos[value.length - 1].components = disassemble(value[value.length - 1]);
       // 한글을 뺀 경우 (길이 변화 X)
-      if (lastComponent === recentChar) {
-        typingInfos.current[value.length - 1].components.pop();
-        setRecentChar(typingInfos.current[value.length - 1].components.at(-1) || '');
-
-        // Backspace시 타속 - 2
-        setTotalTypingCount((prev) => (prev - 2 >= 0 ? prev - 2 : 0));
-        currTypingCount.current = 0;
-
-        if (contentInfos.current[value.length - 1].char === value[value.length - 1]) {
+      if (prevComponents > currComponents) {
+        if (contentInfos[value.length - 1].char === value[value.length - 1]) {
           setTypingStates(`${typingStates.slice(0, -2)}cf`);
-        } else if (
-          typingInfos.current[value.length - 1].components.every(
-            (component, i) => component === contentInfos.current[value.length - 1].components[i],
-          )
-        ) {
-          setTypingStates(`${typingStates.slice(0, -2)}uf`);
+          const count = typingInfos[value.length - 1].components.length;
+          setTypingCount(typingCount + count);
         } else {
           setTypingStates(`${typingStates.slice(0, -2)}if`);
-          currTypingCount.current = 0;
         }
       }
       // 한글을 더한 경우 (길이 변화 X)
       else {
-        typingInfos.current[value.length - 1].components.push(recentChar);
-
-        if (contentInfos.current[value.length - 1].char === value[value.length - 1]) {
+        if (contentInfos[value.length - 1].char === value[value.length - 1]) {
           setTypingStates(`${typingStates.slice(0, -2)}cf`);
-          const tempTypingCount = currTypingCount.current;
-          setTotalTypingCount((prev) => prev + tempTypingCount);
-          currTypingCount.current = 0;
-        } else if (
-          typingInfos.current[value.length - 1].components.every(
-            (component, i) => component === contentInfos.current[value.length - 1].components[i],
-          )
-        ) {
-          setTypingStates(`${typingStates.slice(0, -2)}uf`);
+          const count = typingInfos[value.length - 1].components.length;
+          setTypingCount(typingCount + count);
         } else {
           setTypingStates(`${typingStates.slice(0, -2)}if`);
-          currTypingCount.current = 0;
         }
       }
     }
     // 타이핑하여 글자가 증가한 경우
     else if (value.length > currLength) {
-      typingInfos.current[value.length - 1].char = value[value.length - 1];
-      typingInfos.current[value.length - 1].type = getCharType(value[value.length - 1]);
-      typingInfos.current[value.length - 1].components = [recentChar];
+      typingInfos[value.length - 1].char = value[value.length - 1];
+      typingInfos[value.length - 1].type = getCharType(value[value.length - 1]);
+      typingInfos[value.length - 1].components = disassemble(value[value.length - 1]);
```

Comment:
```suggestion
      typingInfos[value.length - 1] = {
      char : value[value.length - 1],
      type : getCharType(value[value.length - 1]),
      components : disassemble(value[value.length - 1])
      }
```

이와 같은 방식은 어떨까요?
직접 변경하는 방식 보다는 참조를 옮기는 방식이 좋은것으로 알고있습니다.

Reply:
이 부분에 대해서 모두 useRef로 바꾼 상태 입니다.
상태가 아닌 useRef로 선언한 변수는 직접 변경하는 것이 맞겠죠?

Reply:
앗 아니요 이건 state 관련 컨벤션이 아니고, 
자바스크립트에서 객체의 값을 변경할떄는 이런 방식을 사용해야하는 것으로 알고있는데, 정확한 이유를 아직 못찾습니다. ㅜ 
찾아보고 슬랙에 올릴게요!

Reply:
@msdio 해람님은 어떻게 생각하시나요. 혹시 이유를 아시나요?

Reply:
윤우님이 사용하신 방식은 `dot notation`이라고 부르고,
수미님이 제안하신 방식은 `bracket notation`이라고 불러요.

bracket notation 방식이 띄어쓰기 같은 특수문자나 숫자, 변수를 사용할 수 있어서 좀 더 자유롭다고는 알고있는데, dot notation을 사용하는 것을 지양해야 하는 이유는 저도 잘 모르겠습니다...🙄

Reply:
드디어 제가 그렇게 생각한 이유를 알아냈는데, 리액트 불변성에 대한 내용이였어요. 
useState같은 경우에는 이 불변성을 지켜야 하는거 같은데, useRef를 사용할때는 리렌더링과 관계 없어 크게 상관없는것 같네요. 

useState에서는 참조값이 바뀌어야 객체가 변했다고 감지하고, 리렌더링을 시키게 되는데
useState를 사용할때 이처럼 사용하면 참조가 변하지 않았는데 객체 안의 값을 함부로 바꾸면 예상치 못한 버그가 발생할수 있고, 객체가 변경되었다고 감지 하지 못할수도 있다는 문제가 있어 새로운 객체를 생성해서 넣어주도록 하는 방식을 사용해야합니다. 

그런데 위 내용과 제 얕은 지식이 충돌해서.. 헷갈렸네요 😭

[리액트 불변성이란? ](https://hsp0418.tistory.com/171)